### DISCUSSION TITLE
How to Import & Export without downtime

### DISCUSSION THREAD

**Discussion by Christiaan-Mathu** [keycloak/keycloak#31044] [Q&A]
Hi everyone.

I'm seeking the best method to import and export a realm (including users, not just partial) without stopping the Keycloak server. 

Here's what I've found and tried so far:
  1. **Official documentation:** I've reviewed the [export/import documentation](https://www.keycloak.org/server/importExport), which states that "Your Keycloak server instance must not be started when invoking this command."
  
  2. **Exporting while server is running:** In a dev environment, I attempted to run the export command without stopping Keycloak. The realm seemed to export fine, but I encountered errors afterwards because the server couldn't start in 'import/export' mode (as it was already running).
  
  3. **Database backup:** Another approach I've tested is using a tool like pgAdmin for a DB backup. This method works, but I'm unsure of potential unforeseen consequences.
  
  4. **Third-party extension:** I've come across [this extension](https://github.com/cloudtrust/keycloak-export) designed for this use case, but it's no longer maintained.

My main concern is finding a reliable method that doesn't risk data integrity or cause unexpected issues. 

- Has anyone successfully exported/imported users without stopping the server? 
- What's the recommended approach in a production environment where minimizing downtime is crucial?

Any insights or best practices would be greatly appreciated. Thanks in advance for your help!

**Comment by Christiaan-Mathu** [comment]
Hi @jonkoops,

I hope you're doing well! I just wanted to ask if you had any thoughts or opinions on the matter. Your insights would be greatly appreciated.

Thank you in advance!

**Reply by lferrazzini95** [reply]
Hi @Christiaan-Mathu 
Did you find a solution to this issue? We are facing the same limitation and I cannot imagine that this does not impact a lot of people.
Thanks for your answer!

**Reply by Christiaan-Mathu** [reply]
Unfortunately the best solution that I found was just to use the step 2 in my original post. 
As mentioned this is not supported but it was the best option that we had at the time.

I don't personally work with it anymore so you can have a look again to see if they have changed this since.

Maybe try their slack channel, there are a lot of helpful people there as well.

**Reply by shawkins** [reply]
There are several issues here. The primary concern has been consistency, so the recommendation has been to shutdown down the cluster. The hope is that #37991 will aliviate that concern for most scenarios. Given that assumption, the next issue is that import/export commands aren't specifically tweaked to run along side a running server - since they are effectively a server launch that terminates after the import / export operation there can be conflicts. The current workaround is to run the import / export command in a separate container or Pod.
